services/vsync/fallback/vsync_provider_fallback_impl.cc

```diff
@@ -0,0 +1,64 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/services/vsync/fallback/vsync_provider_fallback_impl.h"
+
+#include "base/bind.h"
+#include "base/bind_helpers.h"
+#include "base/message_loop/message_loop.h"
+
+namespace sky {
+namespace services {
+namespace vsync {
+
+VsyncProviderFallbackImpl::VsyncProviderFallbackImpl(
+    mojo::InterfaceRequest<::vsync::VSyncProvider> request)
+    : binding_(this, request.Pass()),
+      phase_(base::TimeTicks::Now()),
+      armed_(false),
+      weak_factory_(this) {}
+
+VsyncProviderFallbackImpl::~VsyncProviderFallbackImpl() = default;
+
+void VsyncProviderFallbackImpl::AwaitVSync(const AwaitVSyncCallback& callback) {
+  pending_.emplace_back(std::move(callback));
+  ArmIfNecessary();
+}
+
+void VsyncProviderFallbackImpl::ArmIfNecessary() {
+  if (armed_) {
+    return;
+  }
+
+  armed_ = true;
+
+  const base::TimeDelta interval = base::TimeDelta::FromSecondsD(1.0 / 60.0);
+
+  const base::TimeTicks now = base::TimeTicks::Now();
+  const base::TimeTicks next = now.SnappedToNextTick(phase_, interval);
```

Comment:
iOS, Android & Mac all have actual VSync providers. So the only platform that might have to use the fallback provider is Linux Desktop.

Though I agree this is a suboptimal solution, it does not accumulate drift. I believe the old solution would schedule the next `BeginFrame` a full frame interval after the `RequestFrame` (and not account for any portion of the time slice already consumed before the `RequestFrame`). What would be a higher quality algorithm without an actual VSync provider?